Applications/Spire/Include/Spire/Ui/NumericFilterPanel.hpp

```diff
@@ -0,0 +1,274 @@
+       * @param parent The parent widget that shows the panel.
+       */
+      NumericFilterPanel(std::shared_ptr<OptionalNumericModel> min_model,
+        std::shared_ptr<OptionalNumericModel> max_model,
+        boost::optional<Type> default_min_value,
+        boost::optional<Type> default_max_value,
+        QString title, QWidget* parent = nullptr);
+
+      /** Returns the model of the minimum value. */
+      const std::shared_ptr<OptionalNumericModel>& get_min_model() const;
+
+      /** Returns the model of the maximum value. */
+      const std::shared_ptr<OptionalNumericModel>& get_max_model() const;
+
+      /** Returns the default minimum value. */
+      boost::optional<Type> get_default_min_value() const;
+
+      /** Returns the default maximum value. */
+      boost::optional<Type> get_default_max_value() const;
+
+      /** Sets the default minimum and maximum value. */
+      void set_default_value(const boost::optional<Type>& default_min,
+        const boost::optional<Type>& default_max);
+
+      /** Connects a slot to the filter signal. */
+      boost::signals2::connection connect_filter_signal(
+        typename const FilterSignal::slot_type& slot) const;
+
+    protected:
+      bool event(QEvent* event) override;
+
+    private:
+      mutable FilterSignal m_filter_signal;
+      std::shared_ptr<OptionalNumericModel> m_min_model;
+      std::shared_ptr<OptionalNumericModel> m_max_model;
+      boost::optional<Type> m_default_min_value;
+      boost::optional<Type> m_default_max_value;
+      boost::optional<Type> m_min_value;
+      boost::optional<Type> m_max_value;
+      FilterPanel* m_filter_panel;
+
+      NumericBox* make_input_field(std::shared_ptr<OptionalNumericModel> model);
+      QHBoxLayout* make_row_layout(QString label_name, NumericBox* input_field);
+      void reset();
+  };
+
+  template<typename T>
+  NumericFilterPanel<T>::NumericFilterPanel(
+    boost::optional<Type> default_min_value,
+    boost::optional<Type> default_max_value,
+    QString tilte, QWidget* parent)
+    : NumericFilterPanel(std::make_shared<LocalOptionalNumericModel>(),
+      std::make_shared<LocalOptionalNumericModel>(), default_min_value,
+      default_max_value, title, parent) {}
+
+  template<typename T>
+  NumericFilterPanel<T>::NumericFilterPanel(
+      std::shared_ptr<OptionalNumericModel> min_model,
+      std::shared_ptr<OptionalNumericModel> max_model,
+      boost::optional<Type> default_min_value,
+      boost::optional<Type> default_max_value,
+      QString title, QWidget* parent)
+      : QWidget(parent),
+        m_min_model(std::move(min_model)),
+        m_max_model(std::move(max_model)),
+        m_default_min_value(std::move(default_min_value)),
+        m_default_max_value(std::move(default_max_value)) {
+    if(m_default_min_value && m_default_max_value &&
+        *m_default_min_value > *m_default_max_value) {
+      std::swap(m_default_min_value, m_default_max_value);
```

Comment:
Is it always the case that these should be swapped, instead of an error being raised? It seems that if this condition is true then there's an error somewhere.

Reply:
I didn't see any errors from here. The VS output window may display an exception which is raised by the parse_decimal function in UiViewer which tries to convert a string to a decimal when the default min/max value is empty, but this output is not caused by the swap.

Reply:
As a follow up, do not mask or cover up for input or programming errors. If a mistake is made try to crash immediately. This promotes a type of local reasoning where errors in the program can be identified within a constant amount of time instead of having to reason globally about an application to track down a problem.